### ISSUE TITLE
Streaming interleave: over + val_split still puts the same row in train and val

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#702]
Found while reviewing #701, which fixes the same defect on the two **eager** interleave paths. Filed separately because the streaming path needs a different mechanism, which is a fair reason to defer it rather than block that PR.

## What happens

`data.streaming: true` with `interleave: over` puts the same row in both the training and the validation split.

Measured on `019c324`, and identically on #701's branch:

```
streaming: true, A=100 B=10, strategy=over, val_split=0.1
  val rows that also appear in train: 10 / 20
  CONTROL probs / concat / under:      0
```

## Mechanism

`interleave_datasets(..., stopping_strategy="all_exhausted")` recycles the shorter stream to exhaust the longer one, so the materialised result contains duplicates of the smaller source. `_finalize` at `loader.py:742` then takes a **positional tail slice** for `val_split`, and a row and its recycled copy are separate entries, so the slice can put one on each side.

Structurally identical to #680. The difference is where the duplication happens — `_cycle_to` on the eager path, `all_exhausted` inside `datasets` on the streaming one.

## Why #701's fix does not transfer

#701 carves `val_split` out **per source before** padding. That works because the eager path has the rows in hand and can count them. A streaming source is not countable ahead of time, so there is nothing to take a fraction of before interleaving starts. The fix needs a different shape — reserving a val stream per source before `interleave_datasets`, tracking which rows have been emitted, or refusing the combination.

## Severity

Silent, and the failure looks like success: validation loss on the recycled source measures memorisation, so the run reports a suspiciously good number rather than an error. Anyone using `data.streaming: true` with `interleave: over` and a non-zero `val_split` is affected. `probs`, `concat` and `under` are clean on this path.

There is also no escape hatch: `DataConfig` has no `data.val` field (`[f for f in DataConfig.model_fields if 'val' in f]` -> `['val_split', 'eval_on_each_dataset']`), so an independently supplied validation set is not available for local interleave. The only workaround today is avoiding `over` on the streaming path.

## Documentation is currently misleading about this

`docs/data.md` opens the strategy table with *"The strategy names mean the same thing on the streaming path as on the local path"*, and after #701 lands the paragraph below it will describe a fix scoped to the eager path. A reader will reasonably conclude streaming is covered. #701 is adding a clause naming this exclusion; this issue is the thing that clause points at.

## Acceptance criteria

- [ ] `streaming: true` + `over` + `val_split > 0` produces train and val sets with **zero** row overlap
- [ ] The test asserts set-disjointness **and** that val is non-empty — a disjointness assertion alone is trivially true when val is empty, which is a real gap found in #701's own tests
- [ ] `probs`, `concat` and `under` on the streaming path are unchanged, with a control test
- [ ] If the chosen answer is to refuse the combination rather than fix it, the refusal happens at config-parse time with a message naming `streaming`, `over` and `val_split` — not several minutes into a run
- [ ] `docs/data.md` no longer implies the streaming path is covered by #701's fix

Reported mechanism and the eager-path fix are @AmirF194's in #680 and #701; the streaming gap was measured during that review.

**Comment by MakazhanAlpamys** [comment]
@v01dst — a courtesy note, not a problem.

I named **#702** to you twice as available: on #679 at 2026-09-05T18:36 and again in my reply after you closed that PR. I did not put an expiry on it, which is the same omission that has now produced four claim races in four days here.

**@dchaudhari7177 has opened #729 against it.** They did nothing wrong — the issue had no assignee, no claim comment, and a `help wanted` label — and the PR reproduces the exact shape from the issue before changing anything:

```
BEFORE  train=180 val=20  overlapping val rows=10
AFTER   train=189 val=11  overlapping val rows=0
```

Ten of twenty, matching the measurement I put in the issue.

So #702 is going ahead with them unless you tell me you had started on it, in which case say so today and I will sort it out with them the way I am doing on #716. I am not holding it open on a maybe — that is the failure mode, not the fix.

Nothing else changes: your `isinstance(messages, list)` guard is merged in `06773c5`, credited in the commit body, the changelog fragment, and `CONTRIBUTORS.md` under **Mahmoud Ahmed** rather than your login, since `gh api users/v01dst` gives the real name and the rest of that file uses one. **#699** is still yours until 2026-09-06 18:00 UTC — that has not expired yet, and I mis-stated it in the first version of this comment. After it, it passes to @swalla02 as announced.

Thank you again for how you handled #679. It remains the best thing anyone has done in this repo this week.

**Comment by MakazhanAlpamys** [comment]
Fixed by @dchaudhari7177 in #729, merged as `d0f6981d`.

Measured against real `datasets` 5.0.1 end to end: on `main`, `over` gives train=179 / val=20 with **10 rows on both sides**; after the fix, train=188 / val=11 with **0 overlap**, and `concat` / `under` / `probs` are untouched. The val set is now sized over distinct rows rather than positions.

Two things beyond the report. A first attempt at the fix silently dropped a genuinely repeated source row — @AmirF194 caught that, and the merged version instead withholds the extra copies **and prints the count**, so the lossy case is loud. And a `val_split` that would leave zero training rows is now refused by name rather than producing an empty train set.

Verified before merge: disabling the `over` carve-out reproduces the leak; gating on `val_split` alone (which would reroute the other three strategies) fails all three control tests — a mutation that passed 65 tests on an earlier head.

Worth knowing: the val set **shrinks** on `over` (20 to 11 on this issue's shape), because the fraction is now taken over distinct rows rather than materialised positions.

Try:

```
soup train --config your.yaml   # with data.interleave strategy: over and a val_split
```
